vpr/src/device/rr_graph/rr_graph_obj.cpp

```diff
@@ -0,0 +1,1753 @@
+    VTR_ASSERT_SAFE(valid_switch_id(switch_id));
+    return switches_[switch_id];
+}
+
+size_t RRGraph::segment_index(RRSegmentId segment_id) const {
+    VTR_ASSERT_SAFE(valid_segment_id(segment_id));
+    return size_t(segment_id);
+}
+
+/*
+ * Get a segment from the segment list with a given id  
+ */
+const t_segment_inf& RRGraph::get_segment(RRSegmentId segment_id) const {
+    VTR_ASSERT_SAFE(valid_segment_id(segment_id));
+    return segments_[segment_id];
+}
+
+RREdgeId RRGraph::find_edge(RRNodeId src_node, RRNodeId sink_node) const {
+    for (auto edge : node_out_edges(src_node)) {
+        if (edge_sink_node(edge) == sink_node) {
+            return edge;
+        }
+    }
+
+    //Not found
+    return RREdgeId::INVALID();
+}
+
+RRNodeId RRGraph::find_node(short x, short y, t_rr_type type, int ptc, e_side side) const {
+    if (!valid_fast_node_lookup()) {
+        build_fast_node_lookup();
+    }
+    size_t itype = type;
+    size_t iside = side;
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((x < 0)                                     /* See if x is smaller than the index of first element */
+        || (size_t(x) > node_lookup_.size() - 1)) { /* See if x is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((y < 0)                                        /* See if y is smaller than the index of first element */
+        || (size_t(y) > node_lookup_[x].size() - 1)) { /* See if y is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    /* itype is always larger than -1, we can skip checking */
+    if (itype > node_lookup_[x][y].size() - 1) { /* See if type is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((ptc < 0)                                                 /* See if ptc is smaller than the index of first element */
+        || (size_t(ptc) > node_lookup_[x][y][type].size() - 1)) { /* See if ptc is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    /* iside is always larger than -1, we can skip checking */
+    if (iside > node_lookup_[x][y][type][ptc].size() - 1) { /* See if side is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    return node_lookup_[x][y][itype][ptc][iside];
+}
+
+RRGraph::node_range RRGraph::find_nodes(short x, short y, t_rr_type type, int ptc) const {
+    if (!valid_fast_node_lookup()) {
+        build_fast_node_lookup();
```

Comment:
Since this sequence shows up so often we could consider wrapping:
```cpp
    if (!valid_fast_node_lookup()) {
        build_fast_node_lookup();
    }
```
into a private member function like `initialize_fast_node_lookup()`

Reply:
Done